src/tokenserver/db/models.rs

```diff
@@ -61,152 +70,294 @@ impl TokenserverDb {
-            .load::<results::GetUser>(&self.inner.conn)?;
+        let mut raw_users = diesel::sql_query(QUERY)
+            .bind::<Text, _>(&params.email)
+            .bind::<Integer, _>(params.service_id)
+            .load::<results::GetRawUser>(&self.inner.conn)?;
 
-        if user_records.is_empty() {
+        if raw_users.is_empty() {
             return Err(DbErrorKind::TokenserverUserNotFound.into());
         }
 
-        user_records.sort_by_key(|user_record| (user_record.generation, user_record.created_at));
-        let user_record = user_records[0].clone();
+        raw_users.sort_by_key(|raw_user| (raw_user.generation, raw_user.created_at));
+
+        // The user with the greatest `generation` and `created_at` is the current user
+        let raw_user = raw_users[0].clone();
+
+        // Collect any old client states that differ from the current client state
+        let old_client_states = raw_users[1..]
+            .iter()
+            .map(|user| user.client_state.clone())
+            .filter(|client_state| client_state != &raw_user.client_state)
+            .collect();
+
+        // Make sure every old row is marked as replaced. They might not be, due to races in row
+        // creation.
+        for old_user in &raw_users[1..] {
+            if old_user.replaced_at.is_none() {
+                let params = params::ReplaceUser {
+                    uid: old_user.uid,
+                    service_id: params.service_id,
+                    replaced_at: old_user.created_at,
+                };
+
+                self.replace_user_sync(params)?;
+            }
+        }
 
-        Ok(user_record)
-    }
+        let user = results::GetUser {
+            uid: raw_user.uid,
+            client_state: raw_user.client_state,
+            generation: raw_user.generation,
+            node: raw_user.node,
+            keys_changed_at: raw_user.keys_changed_at,
+            created_at: raw_user.created_at,
+            old_client_states,
+        };
 
-    #[cfg(test)]
-    fn post_node_sync(&self, node: params::PostNode) -> DbResult<results::PostNode> {
-        let query = r#"
-            INSERT INTO nodes (service, node, available, current_load, capacity, downed, backoff)
-            VALUES (?, ?, ?, ?, ?, ?, ?)
-        "#;
-        diesel::sql_query(query)
-            .bind::<Integer, _>(node.service_id)
-            .bind::<Text, _>(&node.node)
-            .bind::<Integer, _>(node.available)
-            .bind::<Integer, _>(node.current_load)
-            .bind::<Integer, _>(node.capacity)
-            .bind::<Integer, _>(node.downed)
-            .bind::<Integer, _>(node.backoff)
-            .execute(&self.inner.conn)?;
+        Ok(user)
+    }
 
-        let query = r#"
-            SELECT id
-              FROM nodes
+    /// Mark users matching the given email and service ID as replaced.
+    fn replace_users_sync(&self, params: params::ReplaceUsers) -> DbResult<results::ReplaceUsers> {
```

Comment:
Analogous legacy Tokenserver DB method is [here](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/assignment/sqlnode/sql.py#L516)